File: Data Structures & Algorithms/partition-equal-subset-sum/submission-3.cpp
class Solution
{
public:
    vector<vector<int>> memo;
    bool canPartition(vector<int> &nums)
    {
        int sum = 0;
        for (int num : nums)
        {
            sum += num;
        }
        if (sum % 2 == 1)
        {
            return false;
        }
        int half = sum / 2;
        memo.resize(nums.size(), vector<int>(half + 1, -1));
        return dfs(nums, 0, half);
    }

    bool dfs(vector<int> &nums, int i, int half)
    {
        if (i == nums.size())
        {
            return half == 0;
        }
        if (half < 0) {
            return false;
        }
        if (memo[i][half] != -1)
        {
            return memo[i][half];
        }
        memo[i][half] = dfs(nums, i + 1, half) || dfs(nums, i + 1, half - nums[i]);
        return memo[i][half];
    }
};

File: Data Structures & Algorithms/house-robber/submission-3.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        '''
        # recursion method
        def dfs(i):
            if i >= len(nums):
                return 0
            
            return max(nums[i] + dfs(i + 2), dfs(i + 1))
        return max(dfs(0), dfs(1))
        '''

        # botton-up

        memo = [-1] * len(nums)
        def dfs(i):
            if i >= len(nums):
                return 0

            if memo[i] != -1:
                return memo[i]
            
            memo[i] = max(nums[i] + dfs(i + 2), dfs(i + 1))
            return memo[i]
        return max(dfs(0), dfs(1))